---
output: github_document
---
# `QuadST: A Powerful and Robust Approach for Identifying Cellâ€“Cell Interaction-Changed Genes on Spatially Resolved Transcriptomics`

A full description of the method can be found in our [paper](https://doi.org/10.1101/2023.12.04.570019)

## Installation
With R, users can install the MiXcan package directly from GitHub with devtools:

## Example of use
```{r}
install.packages("devtools")
devtools::install_github("songxiaoyu/MiXcan/Package")
```

Below demonstrates the QuadST analysis pipeline for a single cell type pair. Multiple cell type pairs can be analyzed in parallel.

```{r echo = T, results = 'hide'}
library(QuadST)
library(tidyverse)
library(SingleCellExperiment)
library(spatstat)
library(quantreg)
library(QRank)
library(ACAT)
```

### QuadST analysis pipeline
Step 1: Create an anchor-neighbor integrated matrix.
```{r}
data("seqFISHplus_scran_sce")
cell_id = "cellID"
cell_coord1 = "x"
cell_coord2 = "y"
cell_type = "cellClass"
anchor = "Excitatory neuron"
neighbor = "Excitatory neuron"
covariate = "FOV"

sce_an <- create_cellpair_matrix(seqFISHplus_scran_sce, cell_id, cell_coord1, cell_coord2, cell_type, anchor, neighbor, cov=covariate)
sce_an
```

Step 2: Model and test distance-expression association
```{r}
# A. Determine the number of quantile levels, e.g., to ensure that there are at least 5 samples in each quantile level.
anchor_cell_count <- length(colData(sce_an)[, cell_id])
dist_taus <- create_quantile_levels(min_sample_per_quantile = 5, cell_count = anchor_cell_count, max_default = 49)

# B. Subset genes to analyze, e.g., top 25% highly expressed genes.
xm <- apply(t(assay(sce_an, "adjusted.counts")), 2, mean)
xmq <- quantile(xm, 0.75)
gene_to_keep <- names(xm)[xm > xmq]
sce_an_sub <- sce_an[gene_to_keep,]

# C. Transform cell-specific bias adjusted counts to normally distributed values.
assay(sce_an_sub, "normalized_counts") <- transform_count_to_normal(assay(sce_an_sub, "adjusted.counts"))

# D. Provide the colunm names of sce_an_sub that correspond to distance, expression values, and covariates to be used for analysis.
dist <- "distance"
expr <- "normalized_counts"
covariate <- "FOV"
QRpvalue <- test_QuadST_model(sce_an_sub, dist, expr, cov = covariate, tau = dist_taus)
str(QRpvalue)
```

Step 3: Identify ICGs.
```{r}
res <- identify_ICGs(sce_an_sub, QRpvalue, dist, expr, cov = covariate, tau = dist_taus, p_thres = 0.05, fdr = 0.1, ABconst = 0.1)

# A. Check ICGs
str(res$ICGs)

# B. Interaction quantile and distance
str(res$q_int)
str(res$dist_int)
```
